Fix #13156: Introduce new attribute thumbnail_size_in_bytes for Subtopics

srikanthkb opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #13156
2. This PR does the following: 
- Introduce a new attribute thumbnail_size_in_bytes for `Subtopic` domain class.
- Update the class methods (`from_dict()`, `to_dict()`) for `Subtopic` domain class to include the new attribute.
- Code changes for update-flow of `thumbnail_filename` to update `thumbnail_size_in_bytes` for subtopics.
- Performs schema migration for Subtopics

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes made in this PR work correctly.
If the changes in your PRs are autogenerated via a script and you cannot provide proof for the changes then please leave a comment "No proof of changes needed because {{Reason}}".
-->
Creation of a subtopic is successful:
<img width="1137" alt="Screen Shot 2021-07-07 at 2 00 03 PM" src="https://user-images.githubusercontent.com/36214903/124726933-d8decd00-df2b-11eb-9f07-81bd1555dbac.png">
Editing of Subtopic is successful:
<img width="1274" alt="Screen Shot 2021-07-07 at 2 00 50 PM" src="https://user-images.githubusercontent.com/36214903/124727003-ee53f700-df2b-11eb-9591-adb90c9ec1b3.png">

Updating thumbnail of Subtopic is successful:
<img width="1175" alt="Screen Shot 2021-07-07 at 2 02 34 PM" src="https://user-images.githubusercontent.com/36214903/124727193-1e9b9580-df2c-11eb-903e-d18fc746147e.png">

Thumbnail updated successfully:
<img width="1241" alt="Screen Shot 2021-07-07 at 2 03 08 PM" src="https://user-images.githubusercontent.com/36214903/124727237-278c6700-df2c-11eb-830d-4326de9d1e1b.png">

Subtopic saved as dict in TopicModel (with thumbnail_size_in_bytes attribute):
<img width="1280" alt="Screen Shot 2021-07-07 at 2 05 12 PM" src="https://user-images.githubusercontent.com/36214903/124727566-776b2e00-df2c-11eb-87f6-72026f864efd.png">

Subtopic with schema version 3 (in datastore) -- Before migration:
![image](https://user-images.githubusercontent.com/36214903/125049837-04de8780-e0bf-11eb-947c-3264f3355efe.png)

Migration job output:
![image](https://user-images.githubusercontent.com/36214903/125049884-13c53a00-e0bf-11eb-9542-99b447c5713f.png)

Subtopic migrated to version 4 -- After successful migration
![image](https://user-images.githubusercontent.com/36214903/125050017-33f4f900-e0bf-11eb-8c96-4636d58bce31.png)

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

srikanthkb committed:
initial commit

srikanthkb committed:
updated schema migration for subtopics

srikanthkb committed:
added one-off job for updating subtopic thumbnails

srikanthkb committed:
lint fix

srikanthkb committed:
Merge remote-tracking branch 'upstream/develop' into subtopic-schema-migration

srikanthkb committed:
add thumbnail_size_in_bytes argument for tests

srikanthkb committed:
include topic_id as argument for conversion_fn

srikanthkb committed:
removed one-off job for updating subtopic thumbnail size

* Not required -- as the conversion function is directly
  updating the right value for the thumbnail

srikanthkb committed:
lint fix

srikanthkb committed:
fix backend coverage tests

srikanthkb committed:
update thumbnail_size_in_bytes in topic_domain

* use fs_domain isFile api to check if file exists

srikanthkb committed:
Merge remote-tracking branch 'upstream/develop' into subtopic-schema-migration

srikanthkb committed:
update conversion function

srikanthkb committed:
lint fix

srikanthkb committed:
fix backend topic_services_test

srikanthkb committed:
fix skill_services_test failures

srikanthkb committed:
Merge remote-tracking branch 'upstream/develop' into subtopic-schema-migration

srikanthkb committed:
fix topic_domain_test

seanlip commented on core/controllers/admin_test.py:
```diff
@@ -330,7 +330,7 @@ def test_regenerate_topic_related_opportunities_action(self):
             topic_domain.Subtopic(
                 1, 'Title', ['skill_id_1'], 'image.svg',
-                constants.ALLOWED_THUMBNAIL_BG_COLORS['subtopic'][0],
+                constants.ALLOWED_THUMBNAIL_BG_COLORS['subtopic'][0], 0,
```
Is 0 the right value here -- shouldn't we use a reasonable positive integer? One of the things I'm also wondering is whether we should also disallow a value of 0 when validating a PUT request.

srikanthkb replied:
I agree, thumbnail_size_in_bytes should be None or "non-zero integer".

srikanthkb committed:
fix backend tests

* topic_viewer_tests
* topic_and_skills_dashboard_tests
* subtopic_viewer_test

seanlip commented on core/domain/topic_domain.py:
```diff
@@ -1137,6 +1146,33 @@ def create_default_topic(
+        file_system_class = fs_services.get_entity_file_system_class()
+        fs = fs_domain.AbstractFileSystem(file_system_class(
+            feconf.ENTITY_TYPE_TOPIC, topic_id))
+        filename_prefix = 'thumbnail'
```
Make this generic, as we discussed in the other PR.

srikanthkb replied:
Done, for now -- I will be making the change here as well.
This is being handled in [PR 13175](https://github.com/oppia/oppia/pull/13159#pullrequestreview-698846361). Once, that gets merged I will resolve the conflicts and merge this.

seanlip commented on core/domain/topic_domain.py:
```diff
@@ -1137,6 +1146,33 @@ def create_default_topic(
+        filename_prefix = 'thumbnail'
+        filepath = '%s/%s' % (
+            filename_prefix, subtopic_dict['thumbnail_filename'])
+        if fs.isfile(filepath):
```
I think you can do:

```
subtopic_dict['thumbnail_size_in_bytes'] = (
    len(fs.get(filepath)) if fs.isfile(filepath) else None)
```

srikanthkb replied:
Done

seanlip commented on core/domain/topic_domain.py:
```diff
@@ -1464,6 +1506,14 @@ def update_subtopic_thumbnail_filename(
+        file_system_class = fs_services.get_entity_file_system_class()
+        fs = fs_domain.AbstractFileSystem(file_system_class(
+            feconf.ENTITY_TYPE_TOPIC, self.id))
+        filename_prefix = 'thumbnail'
```
Ditto as above.

srikanthkb replied:
Done

seanlip commented on core/domain/topic_domain.py:
```diff
@@ -1137,6 +1146,33 @@ def create_default_topic(
             feconf.CURRENT_STORY_REFERENCE_SCHEMA_VERSION, '', False, '')
 
+    @classmethod
+    def _convert_subtopic_v3_dict_to_v4_dict(cls, topic_id, subtopic_dict):
```
I don't see the tests covering the migration? You probably need to do some work there -- I noticed the backend tests are failing.

srikanthkb replied:
Done, updated.

seanlip commented:
Note: I suggest fixing the backend tests before requesting a second review (unless you have specific questions for me). I can confirm that the overall approach looks reasonable!

srikanthkb committed:
fix backend tests

* update test files
* added tests for update_subtopic_thumbnail() method

srikanthkb committed:
lint fix

srikanthkb committed:
Merge remote-tracking branch 'upstream/develop' into subtopic-schema-migration

srikanthkb committed:
Merge remote-tracking branch 'upstream/develop' into subtopic-schema-migration

srikanthkb committed:
fix android_e2e_test

srikanthkb committed:
fix topic_viewer tests

srikanthkb committed:
add tests for thumbnail validation and update

srikanthkb committed:
Merge remote-tracking branch 'upstream/develop' into subtopic-schema-migration

seanlip commented on core/domain/topic_domain.py:
```diff
@@ -441,6 +450,11 @@ def validate(self):
             raise utils.ValidationError(
                 'Subtopic thumbnail background color is not specified.')
+        if self.thumbnail_filename is not None and (
+                self.thumbnail_size_in_bytes is 0):
```
Technically this should be `==` not `is`.

srikanthkb replied:
Facing the lint error, when using `!=` instead of `is not None`
![image](https://user-images.githubusercontent.com/36214903/124857133-602e4e00-dfc9-11eb-99a9-e4ba5b3e3619.png)

srikanthkb replied:
Done.

seanlip commented on core/domain/topic_domain.py:
```diff
@@ -1137,6 +1151,30 @@ def create_default_topic(
+    @classmethod
+    def _convert_subtopic_v3_dict_to_v4_dict(cls, topic_id, subtopic_dict):
+        """Converts old Subtopic schema to the modern v4 schema. v4 schema
+        introduces the thumbnail_size_in_bytes field
```
Missing period at end.

srikanthkb replied:
Done.

seanlip reviewed (approved):
Thanks -- LGTM.

seanlip commented:
@sagangwee @nithusha21 @BenHenning @DubeySandeep @aks681 PTAL as codeowners. Note that this is time-sensitive since we need to get it in for the July release. Thanks!

srikanthkb committed:
nit fix

srikanthkb committed:
Merge remote-tracking branch 'upstream/develop' into subtopic-schema-migration

srikanthkb committed:
Merge remote-tracking branch 'upstream/develop' into subtopic-schema-migration

DubeySandeep reviewed (approved):
LGTM for the code owner file `core/controllers/admin_test.py`

BenHenning reviewed (approved):
Thanks @srikanthkb. LGTM for codeowners. Noting that the test image might not look very good on Android, but per current code it should work I think.

BenHenning reviewed (changes requested):
Requesting changes per https://github.com/oppia/oppia/pull/13175#discussion_r666038292.

anandwana001 commented on core/domain/topic_services_test.py:
```diff
@@ -408,6 +411,16 @@ def test_update_subtopic_property(self):
+        fs = fs_domain.AbstractFileSystem(
+            fs_domain.GcsFileSystem(
+                feconf.ENTITY_TYPE_TOPIC, self.TOPIC_ID))
+        fs.commit('thumbnail/image.svg', raw_image, mimetype='image/svg+xml')
```
```suggestion
        fs.commit(
            '%s/image.svg' % (constants.ASSET_TYPE_THUMBNAIL), raw_image,
            mimetype='image/svg+xml')
```

srikanthkb replied:
Thanks for pointing this out, updating now.

srikanthkb replied:
Done.

sagangwee reviewed (approved):
@srikanthkb LGTM for codeowners.

srikanthkb committed:
nit fix

srikanthkb committed:
Merge remote-tracking branch 'upstream/develop' into subtopic-schema-migration

srikanthkb committed:
Merge remote-tracking branch 'upstream/develop' into subtopic-schema-migration

srikanthkb committed:
Merge remote-tracking branch 'upstream/develop' into subtopic-schema-migration

srikanthkb committed:
audit jobs for subtopics

srikanthkb commented on core/domain/topic_jobs_one_off_test.py:
```diff
@@ -199,3 +200,58 @@ def test_migration_job_fails_with_invalid_topic(self):
+
+        # This running without errors indicates that deleted topics are
+        # skipped.
+        self.process_and_flush_pending_mapreduce_tasks()
```
Fixing backend tests now.

srikanthkb replied:
Done.

aks681 reviewed (approved):
Lgtm as codeowner

srikanthkb committed:
add tests for audit jobs

srikanthkb committed:
Merge remote-tracking branch 'upstream/develop' into subtopic-schema-migration

srikanthkb committed:
nit fix

srikanthkb commented:
> Requesting changes per [#13175 (comment)](https://github.com/oppia/oppia/pull/13175#discussion_r666038292).

Done, updated the changes.

srikanthkb commented:
@BenHenning PTAL, I have addressed your comments regarding `android_e2e_config.py`. Requesting your approval ASAP, as we are planning to merge this for July cut release -- which is today.

srikanthkb commented:
Hi @seanlip  
I have successfully verified the subtopic-schema-migration with TopicMigrationOneOffJob and SubtopicThumbnailSizeAuditOneOffJob (with @vojtechjelinek  )-- and validated the outputs for subtopic dicts. The results are as expected.
I have also addressed @BenHenning  comments regarding `android_e2e_config.py`. 
We need to get this merged before @DubeySandeep releases the July cut.

seanlip commented:
@srikanthkb -- Could you please provide more context on exactly how @BenHenning's comment has been addressed? I don't see any updates to core/controllers/android_e2e_config.py in this PR.

srikanthkb commented:
> @srikanthkb -- Could you please provide more context on exactly how @BenHenning's comment has been addressed? I don't see any updates to core/controllers/android_e2e_config.py in this PR.

@seanlip 

In `android_e2e_config.py` we were saving a "dummy" thumbnail image on the filesystem, before updating the thumbnail for the topic (so that the update for thumbnail_filename is successful). However, in #13175 and #13242 -- earlier we saved this thumbnail image separately for Topic and Subtopic respectively. 
But @BenHenning pointed out in [comment](https://github.com/oppia/oppia/pull/13175#pullrequestreview-701825435) that we don't have to be doing this twice in `InitializeAndroidTestDataHandler` `post()` method in `android_e2e_config.py`. Because the asset shares the same entity type, and can be reused.

I requested Ben to unblock the change in #13175 (as it was going to be merged first) and assured him that I will be removing the additional code (for saving the dummy thumbnail on filesystem) for subtopics in #13242 .
That is why, we don't find any changes for `android_e2e_config.py` in this PR -- because it is not required to save the thumbnail again for a subtopic.

Does this help ?

seanlip commented:
When you say you "will be removing the additional code (for saving the dummy thumbnail on filesystem) for subtopics in #13242", I would have expected some deletions in this PR corresponding to the removal. Where are those deletions?

seanlip commented:
Ah, actually, I think I see -- those were for your pending changes for this PR, and you've just removed those. Got it -- I think I can approve this on behalf of @BenHenning, then. Thanks!

srikanthkb commented on core/controllers/android_e2e_config.py:
```diff
@@ -122,6 +123,18 @@ def post(self):
         topic.add_uncategorized_skill_id(skill_id)
         topic.add_subtopic(1, 'Test Subtopic Title')
 
+        # Save the dummy image to the filesystem to be used as thumbnail for
```
@seanlip 
This is an earlier commit -- which had the changes for saving the thumbnail for subtopics. 
This change is not present in the latest commit. 
With the latest commit, this change is not present -- because we merged #13175 already.
